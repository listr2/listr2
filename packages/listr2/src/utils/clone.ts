import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
