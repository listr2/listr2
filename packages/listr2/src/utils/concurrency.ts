// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
