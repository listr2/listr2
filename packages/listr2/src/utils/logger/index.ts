export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
