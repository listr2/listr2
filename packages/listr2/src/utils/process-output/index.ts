export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
