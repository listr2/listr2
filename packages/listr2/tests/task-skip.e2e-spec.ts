import { Listr } from '@root'
import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

describe('skip a task', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  it('should skip the task from internal call', async() => {
    await new Listr(
      [
        {
          task: async(_, task): Promise<void> => {
            task.skip('skipped')
          }
        }
      ],
      { renderer: 'test' }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA')
  })

  it('should skip the task from skip method', async() => {
    await new Listr(
      [
        {
          skip(): string {
            return 'skipped'
          },
          task: async(): Promise<void> => {}
        }
      ],
      { renderer: 'test' }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ')
  })

  it('skip to enable by context will work properly in serial', async() => {
    await new Listr(
      [
        {
          task: async(ctx, task): Promise<void> => {
            ctx.test = true
            task.skip('skipped')
          }
        },
        {
          enabled: (ctx): boolean => ctx.test,
          task: async(_, task): Promise<void> => {
            task.output = 'enabled'
          }
        }
      ],
      { renderer: 'test', concurrent: false }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc')
  })

  it('skip to enable by context will not work properly in concurrent', async() => {
    await new Listr(
      [
        {
          task: async(ctx, task): Promise<void> => {
            ctx.test = true
            task.skip('skipped')
          }
        },
        {
          enabled: (ctx): boolean => ctx.test,
          task: async(_, task): Promise<void> => {
            task.output = 'enabled'
          }
        }
      ],
      { renderer: 'test', concurrent: true }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6')
  })

  it.each<[boolean | string, string]>([[true, 'Skipped task without a title.']])(
    'should skip the task from async skip method returning either boolean or string',
    async(skip) => {
      await new Listr(
        [
          {
            skip: async(): Promise<boolean | string> => {
              return skip
            },
            task: async(_, task): Promise<void> => {
              task.output = 'This will never execute.'
            }
          }
        ],
        { renderer: 'test' }
      ).run()

      expectProcessOutputToMatchSnapshot(output, 'dKEmraxPfJH3cEj4bHeTsieEALbqcFIR')
    }
  )
})
