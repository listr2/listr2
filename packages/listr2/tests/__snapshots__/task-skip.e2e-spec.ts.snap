// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`skip a task should skip the task from async skip method returning either boolean or string: dKEmraxPfJH3cEj4bHeTsieEALbqcFIR-exit 1`] = `[]`;

exports[`skip a task should skip the task from async skip method returning either boolean or string: dKEmraxPfJH3cEj4bHeTsieEALbqcFIR-stderr 1`] = `[]`;

exports[`skip a task should skip the task from async skip method returning either boolean or string: dKEmraxPfJH3cEj4bHeTsieEALbqcFIR-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"skip":"Skipped task without a title."},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`skip a task should skip the task from internal call: xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA-exit 1`] = `[]`;

exports[`skip a task should skip the task from internal call: xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA-stderr 1`] = `[]`;

exports[`skip a task should skip the task from internal call: xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`skip a task should skip the task from skip method: uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ-exit 1`] = `[]`;

exports[`skip a task should skip the task from skip method: uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ-stderr 1`] = `[]`;

exports[`skip a task should skip the task from skip method: uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`skip a task skip to enable by context will not work properly in concurrent: WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6-exit 1`] = `[]`;

exports[`skip a task skip to enable by context will not work properly in concurrent: WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6-stderr 1`] = `[]`;

exports[`skip a task skip to enable by context will not work properly in concurrent: WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`skip a task skip to enable by context will work properly in serial: CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc-exit 1`] = `[]`;

exports[`skip a task skip to enable by context will work properly in serial: CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc-stderr 1`] = `[]`;

exports[`skip a task skip to enable by context will work properly in serial: CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":true,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"OUTPUT","data":"enabled","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;
